### ISSUE TITLE
"an empty namespace may not be set when a resource name is provided"

### ISSUE THREAD

**Issue by humblec** [kubernetes/kubernetes#49376]
Noticed below error message in logs with latest kube and glusterfs plugin.

```
E0721 12:42:10.108878    6212 glusterfs.go:150] glusterfs: failed to get endpoints pvc-1983190d-6de1-11e7-ad35-c85b7636c232[an empty namespace may not be set when a resource name is provided]

E0721 12:42:10.109019    6212 reconciler.go:367] Could not construct volume information: MountVolume.NewMounter failed for volume "kubernetes.io/glusterfs/d8a5cdc1-6de1-11e7-ad35-c85b7636c232-pvc-1983190d-6de1-11e7-ad35-c85b7636c232" (spec.Name: "pvc-1983190d-6de1-11e7-ad35-c85b7636c232") pod "d8a5cdc1-6de1-11e7-ad35-c85b7636c232" (UID: "d8a5cdc1-6de1-11e7-ad35-c85b7636c232") with: an empty namespace may not be set when a resource name is provided

````
~~~
[root@localhost demo]# kubectl get ep,svc
NAME                           ENDPOINTS                                                    AGE
ep/glusterfs-dynamic-claim11   172.17.0.1:1,192.168.10.100:1,192.168.10.101:1 + 1 more...   36m
ep/kubernetes                  192.168.43.105:6443                                          37m

NAME                            TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
svc/glusterfs-dynamic-claim11   ClusterIP   10.0.0.120   <none>        1/TCP     36m
svc/kubernetes                  ClusterIP   10.0.0.1     <none>        443/TCP   37m
[root@localhost demo]# kubectl get pod
NAME           READY     STATUS    RESTARTS   AGE   
gluster-pod2   1/1       Running   0          31m
[root@localhost demo]#
~~~
The subjected code path is :

https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/glusterfs/glusterfs.go#L150

I failed to get  how come a pod namespace can become empty  or the reason for this error message.

Any thoughts on this ?

**Comment by humblec** [comment]
@kubernetes/sig-storage

**Comment by humblec** [comment]
@rootfs @jsafrane Any thoughts on the possibility of this error ?

**Comment by rootfs** [comment]
Is the `gluster-pod` using the endpoint? Can you post pod yaml and `kubecctl describe pod` ?

**Comment by humblec** [comment]
```
[root@localhost demo]# cat nginx1.yaml 
apiVersion: v1
kind: Pod
metadata: 
  labels: 
    name: gluster-pod2
  name: gluster-pod2
spec: 
  containers: 
    - 
      image: "gcr.io/google_containers/nginx-slim:0.8"
      name: gluster-pod2
      ports: 
        - 
          containerPort: 80
          name: web
      securityContext: 
        privileged: true
      volumeMounts: 
        - 
          mountPath: /usr/share/nginx/html
          name: gluster-vol1
  volumes: 
    - 
      name: gluster-vol1
      persistentVolumeClaim: 
        claimName: claim11
```

@rootfs this pod use claim11 which has `glusterfs-dynamic-claim11` as its endpoint.  I was able to run this pod successfully. So I am in doubt that, this error comes somewhere in pod lifecycle, but wondering in which state it can say POD dont have a namespace.

**Comment by rootfs** [comment]
or maybe the endpoint was not accessible by the time the Pod was created but later became available.

**Comment by humblec** [comment]
hmmmm.. possible, but the error says for *.Get() says "empty" namespace and it kind of saying a 'resource name' is provided.  Its confusing.

**Comment by xiangpengzhao** [comment]
Original error from here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/request.go#L606-L609
```go
// TODO: added to catch programmer errors (invoking operations with an object with an empty namespace)
if (r.verb == "GET" || r.verb == "PUT" || r.verb == "DELETE") && r.namespaceSet && len(r.resourceName) > 0 && len(r.namespace) == 0 {
	return fmt.Errorf("an empty namespace may not be set when a resource name is provided")
}
```

What's the output of `kubecctl describe pod`? What's its namespace?

**Comment by hpgood** [comment]
I have the same problem,and the log is:
`2017/10/16 下午1:14:07E1016 05:14:07.451170    4302 glusterfs.go:148] glusterfs: failed to get endpoints pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0[an empty namespace may not be set when a resource name is provided]
2017/10/16 下午1:14:07E1016 05:14:07.451180    4302 reconciler.go:367] Could not construct volume information: MountVolume.NewMounter failed for volume "kubernetes.io/glusterfs/8e6a9036-b222-11e7-9b9b-02e11a87f365-pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0" (spec.Name: "pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0") pod "8e6a9036-b222-11e7-9b9b-02e11a87f365" (UID: "8e6a9036-b222-11e7-9b9b-02e11a87f365") with: an empty namespace may not be set when a resource name is provided
2017/10/16 下午1:17:07E1016 05:17:07.457846    4302 glusterfs.go:148] glusterfs: failed to get endpoints pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0[an empty namespace may not be set when a resource name is provided]
2017/10/16 下午1:17:07E1016 05:17:07.457879    4302 reconciler.go:367] Could not construct volume information: MountVolume.NewMounter failed for volume "kubernetes.io/glusterfs/20a77095-b22f-11e7-9b9b-02e11a87f365-pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0" (spec.Name: "pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0") pod "20a77095-b22f-11e7-9b9b-02e11a87f365" (UID: "20a77095-b22f-11e7-9b9b-02e11a87f365") with: an empty namespace may not be set when a resource name is provided
2017/10/16 下午1:17:07E1016 05:17:07.458027    4302 glusterfs.go:148] glusterfs: failed to get endpoints pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0[an empty namespace may not be set when a resource name is provided]
2017/10/16 下午1:17:07E1016 05:17:07.458051    4302 reconciler.go:367] Could not construct volume information: MountVolume.NewMounter failed for volume "kubernetes.io/glusterfs/4ee91564-b22f-11e7-9b9b-02e11a87f365-pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0" (spec.Name: "pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0") pod "4ee91564-b22f-11e7-9b9b-02e11a87f365" (UID: "4ee91564-b22f-11e7-9b9b-02e11a87f365") with: an empty namespace may not be set when a resource name is provided
2017/10/16 下午1:17:07E1016 05:17:07.458108    4302 glusterfs.go:148] glusterfs: failed to get endpoints pvc-bd4de7eb-9850-11e7-9dce-02e11aab97b0[an empty namespace may not be set when a resource name is provided]`

**Comment by hpgood** [comment]
I found that:

https://github.com/kubernetes/kubernetes/blob/0ba7c52b8cafce1af39499bee993a2df38640a38/pkg/kubelet/volumemanager/reconciler/reconciler.go#L430-L434

should be:

 namespace:="the real namespace of  volume.podName"
 pod := &v1.Pod{ 
 	ObjectMeta: metav1.ObjectMeta{ 
 		UID: types.UID(volume.podName), 
                Namespace : types.UID(namespace), 
 	}, 
 }

**Comment by hpgood** [comment]
I had done the similar things (clear  the old volume of glusterfs ) by python shell:
kubernetes-tools/findFreePods.py
https://github.com/hpgood/kubernetes-tools/blob/master/findFreePods.py

**Comment by jl-dos** [comment]
This error happens to me as well consistently on mounted volumes while trying to delete. GlusterFS storage backend to pods are running in different namespace than default. Pods hang indefinitely in terminated state until I kill the associated glusterfsd pid and then unmount the gluster volume.

During this time I get errors:
1) "Could not construct volume information: MountVolume.NewMounter failed for volume 'xxxxxx' with: an empty namespace may not be set when a resource name is provided"
2) "reconciler.go:186] operationExecutor.UnmountVolume started for volume 'xxxxxx' "

**Comment by humblec** [comment]
This issue is alreaady fixed in kube >=1.6,  please reopen if you face this issue again. I am closing it for now.
